In [1]:
import pandas as pd

In [2]:
xls = pd.ExcelFile('./attention_to_details.xlsx')
df1 = pd.read_excel(xls, 'Справочник точек')
df2 = pd.read_excel(xls, 'Выручка по обучающей выборке')

/var/folders/s5/t6cm63bn7zn0tzshtl9dsnm80000gn/T/ipykernel_94279/246785079.py:2: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  df1 = pd.read_excel(xls, 'Справочник точек')


# 0. Базовая характеристика датасета

In [3]:
df1.shape, df2.shape

((246, 21), (155, 47))

In [4]:
df1.info()
df1.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246 entries, 0 to 245
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   id точки             245 non-null    float64       
 1   Адрес                246 non-null    object        
 2   Широта               242 non-null    float64       
 3   Долгота              242 non-null    float64       
 4   Дата открытия        245 non-null    object        
 5   Дата закрытия        9 non-null      datetime64[ns]
 6   Площадь магазина     246 non-null    int64         
 7   Тип входа            244 non-null    object        
 8   Площадь ТЦ           204 non-null    float64       
 9   Формат магазина      246 non-null    object        
 10  Парковка             226 non-null    object        
 11  Этаж                 241 non-null    object        
 12  Время работы         232 non-null    object        
 13  Заполненность ТЦ     182 non-null  

id точки      Широта     Долгота  Площадь магазина     Площадь ТЦ  \
count  2.450000e+02  242.000000  242.000000        246.000000     204.000000   
mean   1.132649e+05   47.270235   55.692249        380.837398   69561.509804   
std    4.332589e+05   24.447551    5.587842        206.211319   67128.326340   
min    8.955000e+03   20.500255   43.019324         91.000000     700.000000   
25%    4.201100e+04   33.383124   53.376779        251.000000   19958.500000   
50%    4.230600e+04   37.893216   55.794705        330.000000   47800.000000   
75%    4.260800e+04   50.118003   59.832158        429.750000  103455.750000   
max    2.633948e+06  158.662682   91.758249       1488.000000  396000.000000   

       Заполненность ТЦ  Рядом с эскалатором  Рядом со входом  \
count        182.000000           246.000000       246.000000   
mean           2.472527             0.451220         0.487805   
std            0.858376             0.506749         0.540081   
min            0.000000             0.000000         0.000000   
25%            2.000000             0.000000         0.000000   
50%            3.000000             0.000000         0.000000   
75%            3.000000             1.000000         1.000000   
max            3.000000             2.000000         3.000000   

       Конкуренты в ТЦ  Расположение в ТЦ      Толока        Люкс  
count        94.000000         224.000000  246.000000  246.000000  
mean          0.776596           8.042411   56.703189   69.644964  
std           0.705649           2.911510   35.526801   12.816096  
min           0.000000           0.000000   10.454353   36.724138  
25%           0.000000           6.000000   32.953392   61.974138  
50%           1.000000           8.000000   49.812128   69.287356  
75%           1.000000          10.000000   69.030800   75.689655  
max           3.000000          14.000000  248.234096  124.241379

In [5]:
df2.info()
df2.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 47 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id точки    152 non-null    float64
 1   2015-01-01  116 non-null    float64
 2   2015-02-01  116 non-null    float64
 3   2015-03-01  116 non-null    float64
 4   2015-04-01  116 non-null    float64
 5   2015-05-01  117 non-null    float64
 6   2015-06-01  117 non-null    float64
 7   2015-07-01  117 non-null    float64
 8   2015-08-01  118 non-null    float64
 9   2015-09-01  119 non-null    float64
 10  2015-10-01  119 non-null    float64
 11  2015-11-01  120 non-null    object 
 12  2015-12-01  126 non-null    object 
 13  2016-01-01  127 non-null    object 
 14  2016-02-01  130 non-null    object 
 15  2016-03-01  131 non-null    object 
 16  2016-04-01  131 non-null    object 
 17  2016-05-01  131 non-null    object 
 18  2016-06-01  129 non-null    float64
 19  2016-07-01  132 non-null    f

id точки    2015-01-01    2015-02-01    2015-03-01    2015-04-01  \
count  1.520000e+02  1.160000e+02  1.160000e+02  1.160000e+02  1.160000e+02   
mean   1.252712e+05  3.471735e+05  4.273415e+05  5.195471e+05  3.705748e+05   
std    4.642643e+05  2.718016e+05  3.160264e+05  3.826376e+05  3.064474e+05   
min    1.136600e+04  6.583214e+04  6.918367e+04  1.007257e+05  6.457052e+04   
25%    4.204725e+04  1.961744e+05  2.461139e+05  3.017942e+05  2.043461e+05   
50%    4.231500e+04  2.713618e+05  3.506062e+05  4.295962e+05  3.021539e+05   
75%    4.261525e+04  4.105882e+05  5.079262e+05  5.949454e+05  4.296261e+05   
max    2.634011e+06  2.194019e+06  2.613882e+06  3.197476e+06  2.605320e+06   

         2015-05-01    2015-06-01    2015-07-01    2015-08-01    2015-09-01  \
count  1.170000e+02  1.170000e+02  1.170000e+02  1.180000e+02  1.190000e+02   
mean   3.323632e+05  3.827738e+05  3.470348e+05  3.753261e+05  3.869094e+05   
std    2.589376e+05  3.129426e+05  2.744150e+05  3.032558e+05  3.236095e+05   
min    6.253685e+04  8.388060e+04  5.668160e+04  7.165213e+04  1.613360e+04   
25%    1.856185e+05  2.243160e+05  1.993767e+05  2.144706e+05  2.123922e+05   
50%    2.748710e+05  3.020804e+05  2.743375e+05  2.952706e+05  3.117546e+05   
75%    3.937498e+05  4.388415e+05  3.974613e+05  4.308421e+05  4.518621e+05   
max    2.190041e+06  2.776213e+06  2.454760e+06  2.714818e+06  2.865671e+06   

       ...    2018-01-01    2018-02-01    2018-04-01    2018-05-01  \
count  ...  1.510000e+02  1.510000e+02  1.520000e+02  1.510000e+02   
mean   ...  3.764636e+05  4.305306e+05  3.674662e+05  3.560665e+05   
std    ...  3.229271e+05  3.831301e+05  3.495308e+05  3.431532e+05   
min    ...  1.000000e+00  1.000000e+00  7.328064e+04  6.698472e+04   
25%    ...  2.125980e+05  2.392246e+05  2.037045e+05  1.904644e+05   
50%    ...  3.015974e+05  3.271573e+05  2.745617e+05  2.630475e+05   
75%    ...  4.353563e+05  4.997368e+05  4.197584e+05  4.117321e+05   
max    ...  2.894688e+06  3.514717e+06  3.322619e+06  3.091117e+06   

         2018-06-01    2018-07-01    2018-08-01    2018-09-01    2018-10-01  \
count  1.510000e+02  1.510000e+02  1.530000e+02  1.530000e+02  1.520000e+02   
mean   4.106494e+05  3.581560e+05  3.872170e+05  3.783353e+05  3.881171e+05   
std    4.032322e+05  3.541268e+05  3.729617e+05  3.691852e+05  3.801335e+05   
min    7.219403e+04  5.531162e+04  6.369847e+04  6.523125e+04  7.567381e+04   
25%    2.162632e+05  1.973481e+05  2.143401e+05  2.034994e+05  2.054043e+05   
50%    3.007319e+05  2.654039e+05  2.948969e+05  2.828758e+05  2.970651e+05   
75%    4.589798e+05  4.050429e+05  4.428005e+05  4.365966e+05  4.500905e+05   
max    3.639575e+06  3.160223e+06  3.301408e+06  3.201380e+06  3.307600e+06   

         2018-11-01  
count  1.530000e+02  
mean   4.378152e+05  
std    4.307982e+05  
min    7.354225e+04  
25%    2.260162e+05  
50%    3.329472e+05  
75%    5.117739e+05  
max    3.909895e+06  

[8 rows x 36 columns]

In [6]:
df1.head(10)

id точки                                              Адрес      Широта  \
0    8955.0   Оренбургская обл, Оренбург г, Шарлыкское ш, д...   55.118272   
1   11007.0  Алтайский край, г. Барнаул, ул. Малахова, дом ...   83.696799   
2   11366.0  Амурская область, Благовещенск, улица Мухина, 114  127.510357   
3   11810.0  Архангельская обл, г. Архангельск , Троицкий п...   40.528249   
4   15041.0  Архангельская обл, г. Архангельск, Троицкий пр...   40.526317   
5   15087.0  Архангельская обл, г. Архангельск, ул. Воскрес...   40.535400   
6   15621.0  Архангельская обл, г. Северодвинск, ул. Ломоно...   39.785522   
7   15775.0  Астраханская обл, г. Астрахань, ул. Боевая, до...   48.023172   
8   16102.0   Барнаул г, Ленина пр-кт, дом № 102В, ТЦ "Пионер"   83.764275   
9   16119.0  Башкортостан Респ, Уфа г, Цюрупы ул, дом № 97 ...   55.958263   

     Долгота        Дата открытия Дата закрытия  Площадь магазина  \
0  51.843127  2016-11-09 00:00:00           NaT               478   
1  53.356954  2010-04-26 00:00:00           NaT               332   
2  50.280435  2017-08-25 00:00:00           NaT               356   
3  64.532316  2012-11-03 00:00:00           NaT               317   
4  64.534327  2006-12-09 00:00:00           NaT               176   
5  64.541900  2014-11-20 00:00:00           NaT               416   
6  64.550364  2003-11-09 00:00:00           NaT               178   
7  46.338313  2014-02-05 00:00:00           NaT               521   
8  53.362114  2017-06-07 00:00:00           NaT               369   
9  54.736556  2018-05-23 00:00:00           NaT               662   

                                  Тип входа  Площадь ТЦ Формат магазина  ...  \
0                     без сквозного прохода    130000.0              ТЦ  ...   
1                           сквозной проход     31500.0              ТЦ  ...   
2                     без сквозного прохода     36000.0              ТЦ  ...   
3  отдельный вход с улицы + сквозной проход     35000.0              ТЦ  ...   
4                           сквозной проход     15332.0              ТЦ  ...   
5                     без сквозного прохода     60000.0              ТЦ  ...   
6                           сквозной проход     20000.0              ТЦ  ...   
7                     без сквозного прохода    130000.0              ТЦ  ...   
8  отдельный вход с улицы + сквозной проход     80000.0              ТЦ  ...   
9                           сквозной проход     48000.0              ТЦ  ...   

     Этаж      Время работы Заполненность ТЦ  Рядом с эскалатором  \
0       3  с 10:00 до 22:00              NaN                    0   
1       1  с 10:00 до 22:00              NaN                    0   
2       1  с 10:00 до 22:00              3.0                    0   
3       1  с 10:00 до 21:00              3.0                    0   
4  цоколь  с 10:00 до 21:00              3.0                    1   
5       1  с 10:00 до 21:00              3.0                    1   
6       1  с 10:00 до 21:00              3.0                    0   
7       1  с 10:00 до 22:00              NaN                    0   
8       1  с 10:00 до 22:00              0.0                    0   
9       1  с 10:00 до 22:00              3.0                    1   

   Рядом со входом  Конкуренты в ТЦ  Расположение в ТЦ      Толока       Люкс  \
0                0              1.0                7.0   37.857582  69.500000   
1                1              NaN                9.0   57.810494  66.896552   
2                1              1.0                6.0   49.959831  68.727273   
3                1              0.0               10.0  128.459158  64.517241   
4                0              NaN                0.0  103.199885  78.275862   
5                1              NaN               10.5   71.418989  67.035714   
6                0              0.0                8.0   82.880454  64.034483   
7                1              1.0                8.0   38.935806  62.827586   
8             

### Примечание: 
Некоторые данные имеют тип данных `Object`, хотя могу иметь более простой формат, например, `Этаж` может быть представлен через `int`, выборка может быть представлена в бинарном формате, но поскольку в тестовом задании не требуется обучать модель МО этот этап пропускается. Из метода `info` видно, что в обоих датафреймах отсутствуют данные в некоторых колонках.

# 1. Подготовка данных

- Проверка на отсутствующиеся и повторяющиеся значения
- Определить сколько уникальных значений выборки (убрать повторы)
- Распределение на тестовую и обучающию выборки
- Определить сколько уникальных значений формата магазинов (убрать повторы)
- Определить мини ТЦ, открывшиеся до 2017г (отфильтровать `df2` по этим мини ТЦ)
- Определить все уникальные значения парковок (убрать повторы, отфильтровать бесплатные)

In [7]:
df1.loc[df1.duplicated()]

id точки                                              Адрес     Широта  \
238   42645.0  Респ. Татарстан, г. Нижнекамск, ул. Химиков, д...  51.814376   

       Долгота        Дата открытия Дата закрытия  Площадь магазина  \
238  55.631285  2006-04-16 00:00:00           NaT               255   

                 Тип входа  Площадь ТЦ Формат магазина  ... Этаж  \
238  без сквозного прохода      2887.0         Мини ТЦ  ...    1   

        Время работы Заполненность ТЦ  Рядом с эскалатором  Рядом со входом  \
238  с 9:00 до 21:00              3.0                    1                1   

     Конкуренты в ТЦ  Расположение в ТЦ     Толока      Люкс    Выборка  
238              NaN               10.0  29.981741  54.62069  Обучающая  

[1 rows x 21 columns]

In [8]:
df1["Адрес"].loc[df1.duplicated()], df1["id точки"].loc[df1.duplicated()]

(238    Респ. Татарстан, г. Нижнекамск, ул. Химиков, д...
 Name: Адрес, dtype: object,
 238    42645.0
 Name: id точки, dtype: float64)

В датасете присутствует 1 повторяющийся семпл, удалим его

In [9]:
df1.drop_duplicates()

id точки                                              Адрес      Широта  \
0       8955.0   Оренбургская обл, Оренбург г, Шарлыкское ш, д...   55.118272   
1      11007.0  Алтайский край, г. Барнаул, ул. Малахова, дом ...   83.696799   
2      11366.0  Амурская область, Благовещенск, улица Мухина, 114  127.510357   
3      11810.0  Архангельская обл, г. Архангельск , Троицкий п...   40.528249   
4      15041.0  Архангельская обл, г. Архангельск, Троицкий пр...   40.526317   
..         ...                                                ...         ...   
241  2633929.0  Ямало-Ненецкий АО, г. Новый Уренгой, ул. Сибир...   76.631279   
242  2633930.0  Ямало-Ненецкий АО, Ноябрьск г, Советская ул, д...   75.465427   
243  2633935.0  Ярославская обл, г. Ярославль, ул. Победы, дом...   39.869236   
244  2633946.0  Ярославская обл, г. Ярославль, ул. Свободы, до...   39.883707   
245  2633948.0  Ярославская обл, Ярославский район, Телегински...   39.783545   

       Долгота        Дата открытия Дата закрытия  Площадь магазина  \
0    51.843127  2016-11-09 00:00:00           NaT               478   
1    53.356954  2010-04-26 00:00:00           NaT               332   
2    50.280435  2017-08-25 00:00:00           NaT               356   
3    64.532316  2012-11-03 00:00:00           NaT               317   
4    64.534327  2006-12-09 00:00:00           NaT               176   
..         ...                  ...           ...               ...   
241  66.082833  2014-08-01 00:00:00           NaT               326   
242  63.203380  2016-04-16 00:00:00           NaT               707   
243  57.626872  2013-09-14 00:00:00           NaT               423   
244  57.624998  2008-08-28 00:00:00           NaT               290   
245  57.645195  2006-12-12 00:00:00           NaT               319   

                                    Тип входа  Площадь ТЦ Формат магазина  \
0                       без сквозного прохода    130000.0              ТЦ   
1                             сквозной проход     31500.0              ТЦ   
2                       без сквозного прохода     36000.0              ТЦ   
3    отдельный вход с улицы + сквозной проход     35000.0              ТЦ   
4                             сквозной проход     15332.0              ТЦ   
..                                        ...         ...             ...   
241                     без сквозного прохода     40000.0              ТЦ   
242                     без сквозного прохода         NaN         Мини ТЦ   
243                     без сквозного прохода    120000.0              ТЦ   
244                     без сквозного прохода      4000.0              ТЦ   
245                     без сквозного прохода     42000.0              ТЦ   

     ...    Этаж      Время работы Заполненность ТЦ  Рядом с эскалатором  \
0    ...       3  с 10:00 до 22:00              NaN                    0   
1    ...       1  с 10:00 до 22:00              NaN                    0   
2    ...       1  с 10:00 до 22:00              3.0                    0   
3    ...       1  с 10:00 до 21:00              3.0                    0   
4    ...  цоколь  с 10:00 до 21:00              3.0                    1   
..   ...     ...               ...              ...                  ...   
241  ...       1  с 10:00 до 22:00              3.0                    1   
242  ...       1               NaN              2.0                    0   
243  ...       1  с 10:00 до 22:00              3.0                    0   
244  ...       1  с 10:00 до 21:00              NaN                    0   
245  ...       1  с 10:00 до 22:00              3.0                    1   

     Рядом со входом  Конкуренты в ТЦ  Расположение в ТЦ      Толока  \
0                  0              1.0                7.0   37.857582   
1                  1              NaN                9.0   57.810494   
2                  1              1.0                6.0   49.959831   
3                  1              0.0               10

In [10]:
df1.isna().sum()

id точки                 1
Адрес                    0
Широта                   4
Долгота                  4
Дата открытия            1
Дата закрытия          237
Площадь магазина         0
Тип входа                2
Площадь ТЦ              42
Формат магазина          0
Парковка                20
Этаж                     5
Время работы            14
Заполненность ТЦ        64
Рядом с эскалатором      0
Рядом со входом          0
Конкуренты в ТЦ        152
Расположение в ТЦ       22
Толока                   0
Люкс                     0
Выборка                  1
dtype: int64

Среди недостающих значений присутсвуют, те которые необоходимы для решения задачи - `id` и `Выборка`.
Семпл с отсутсвующем значением `id` необохдимо исключить.

In [11]:
df1[df1['id точки'].isna()]

id точки                                              Адрес     Широта  \
229       NaN  Тульская обл, г. Тула, ул. Советская, дом № 47...  37.618866   

       Долгота        Дата открытия Дата закрытия  Площадь магазина  \
229  54.191774  2010-03-23 00:00:00           NaT               189   

           Тип входа  Площадь ТЦ Формат магазина  ... Этаж      Время работы  \
229  сквозной проход     51000.0         Мини ТЦ  ...    1  с 10:00 до 22:00   

    Заполненность ТЦ  Рядом с эскалатором  Рядом со входом  Конкуренты в ТЦ  \
229              2.0                    0                0              1.0   

     Расположение в ТЦ     Толока       Люкс    Выборка  
229                2.0  49.039977  73.137931  Обучающая  

[1 rows x 21 columns]

In [12]:
df1 = df1[df1['id точки'].notna()]

In [13]:
unique_samples = df1['Выборка'].unique()
unique_samples

array(['Обучающая', 'Тестовая', '-', nan, 'Тест', 'Обyчающая'],
      dtype=object)

### Замечание: 
Колонка с выборкой должна состоять только из 2х значений. Как можно заметить здесь имеются не определенные значения, сокращенные значения и значения, судя по всему, содержащие латиницу и кириллицу в одном слове. 

Можно использовать 2 способа чтобы привести к виду с двумя значениями:
1. Отфильтровать обущающие по `id` из выборки, остальные задать как тестовые.
2. Сначала исправить значения `-` на `nan`, оставить слово обучающая только с кириллицой, тест заменить на тестовая.
Первый способ более практичный, но с ним можно потерять информацию о том были ли в датасети заведомо ложные тестовые значения(заданы как тестовые, но их `id` присутствуют в обучающей выборке), поэтому далее будет рассмотрен второй способ.

Ниже приведем к виду с 3 значениями (обучающая, тестовая, nan)

In [14]:
import re
re.search(r'[a-zA-Z]', unique_samples[-1]) 

<re.Match object; span=(2, 3), match='y'>

В полученном массиве с уникальными значениями в выборке последнее слово начинается с латинской буквы `y` заменим на слово, состоящее полностью из кириллицы

In [15]:
df1.loc[df1['Выборка'] == unique_samples[-1], 'Выборка'] = unique_samples[0]
df1.loc[df1['Выборка'] == unique_samples[2], 'Выборка'] = unique_samples[3]
df1.loc[df1['Выборка'] == unique_samples[4], 'Выборка'] = unique_samples[1]

In [16]:
unique_samples = df1['Выборка'].unique()
unique_samples

array(['Обучающая', 'Тестовая', nan], dtype=object)

Теперь нужно проверить есть ли семплы из `df2` с отсутсвующими `id` и семплы из `df1` со значениями `Тестовая` и `nan` присутствующие в обучающей выборке `df2`

In [17]:
df2[df2['id точки'].isna()]

id точки  2015-01-01  2015-02-01  2015-03-01  2015-04-01  2015-05-01  \
108       NaN         NaN         NaN         NaN         NaN         NaN   
129       NaN         NaN         NaN         NaN         NaN         NaN   
154       NaN         NaN         NaN         NaN         NaN         NaN   

     2015-06-01  2015-07-01  2015-08-01  2015-09-01  ...     2018-01-01  \
108         NaN         NaN         NaN         NaN  ...            NaN   
129         NaN         NaN         NaN         NaN  ...  114585.758339   
154         NaN         NaN         NaN         NaN  ...  502913.053140   

        2018-02-01     2018-04-01     2018-05-01  2018-06-01     2018-07-01  \
108            NaN            NaN            NaN         NaN            NaN   
129  118510.400000   94791.263333   93654.952258  108397.950  100705.532258   
154  594450.393471  560283.324000  514546.421768  644308.088  533981.383006   

        2018-08-01     2018-09-01     2018-10-01  2018-11-01  
108            NaN            NaN            NaN         NaN  
129  116268.127419   97226.428333   97570.266129  121038.385  
154  562082.839742  575194.942000  724658.394194  726229.020  

[3 rows x 47 columns]

In [18]:
df2 = df2[df2['id точки'].notna()]

In [19]:
df1['id точки'] = df1['id точки'].astype(int)
df2['id точки'] = df2['id точки'].astype(int)


In [20]:
intersected_id = list(set(df1['id точки']).intersection(df2['id точки']))

In [21]:
df1.loc[df1['id точки'].isin(intersected_id), 'Выборка'].unique()


array(['Обучающая', 'Тестовая', nan], dtype=object)

### Замечание: 
Обучающая выборка содержит, значения помеченные в `df1` как `Тестовая`, это говорит о не точности предоставленных данных клиентом. Поскольку эти данные представлены в обучающей выборке они должны быть заданы как `Обучающие`. Аналогично сделаем и для `nan`.

In [22]:
df1.loc[df1['id точки'].isin(intersected_id), 'Выборка']

2      Обучающая
3      Обучающая
4       Тестовая
7      Обучающая
8      Обучающая
         ...    
232    Обучающая
233    Обучающая
234    Обучающая
235    Обучающая
236     Тестовая
Name: Выборка, Length: 145, dtype: object

In [23]:
df1.loc[~df1['id точки'].isin(intersected_id), 'Выборка']

0      Обучающая
1      Обучающая
5       Тестовая
6      Обучающая
9       Тестовая
         ...    
241     Тестовая
242    Обучающая
243    Обучающая
244    Обучающая
245    Обучающая
Name: Выборка, Length: 100, dtype: object

In [24]:
df1.loc[df1['id точки'].isin(intersected_id), 'Выборка'] = "Обучающая"

In [25]:
df1['Выборка'].unique()

array(['Обучающая', 'Тестовая', nan], dtype=object)

Оставшиеся значения `nan` и `Обучающая`(`id` который отсутствуют в обучающей выборке) заменим на `Тестовая`, тк они осутствуют в `df2`

In [26]:
df1.loc[~df1['id точки'].isin(intersected_id), 'Выборка'] = "Тестовая"

In [27]:
df1['Выборка'].unique()

array(['Тестовая', 'Обучающая'], dtype=object)

Теперь выборки распределены корректно. Для ответа на первый вопрос необходимо уникальность значений формата магазина аналогично выборке

In [28]:
unique_samples = df1['Формат магазина'].unique()
unique_samples

array(['ТЦ', 'Street', 'Мини ТЦ', 'Стрит', 'Мини-ТЦ', 'Strееt', 'Стрт'],
      dtype=object)

Унифицируем все значения стрит и мини ТЦ в одно, для ответа на 2 первых вопроса

In [29]:
unify_street = [
    unique_samples[1],
    unique_samples[3],
    unique_samples[5],
    unique_samples[6]
]

In [30]:
df1.loc[df1['Формат магазина'].isin(unify_street), 'Формат магазина'] = "Стрит"

In [31]:
df1['Формат магазина'].unique()

array(['ТЦ', 'Стрит', 'Мини ТЦ', 'Мини-ТЦ'], dtype=object)

In [32]:
unify_mall = [
    unique_samples[2],
    unique_samples[4]
]
unify_mall

['Мини ТЦ', 'Мини-ТЦ']

In [33]:
df1.loc[df1['Формат магазина'].isin(unify_mall), 'Формат магазина'] = "Мини ТЦ"

In [34]:
df1['Формат магазина'].unique()

array(['ТЦ', 'Стрит', 'Мини ТЦ'], dtype=object)

## Вопрос 1. Сколько магазинов формата стрит в тестовой выборке?

In [35]:
shop_format = "Стрит"
sample_type = "Тестовая"
pd.set_option('display.max_columns', 30)
street_shops = df1.query('Выборка == @sample_type and `Формат магазина` == @shop_format')
street_shops

id точки                                              Адрес     Широта  \
16      18010  Волгоградская обл, г. Волгоград, ул. Рабоче-Кр...  44.501075   
72      42059          г. Санкт-Петербург, Невский пр., дом № 27  30.327007   
155     42466  Новгородская обл, г. Великий Новгород, ул. Газ...  31.271550   
159     42481  Новосибирская обл, Новосибирск г, Красный пр-к...  82.919389   
166     42505  Пензенская обл, г. Пенза, ул. Московская, дом ...  45.017022   
221     42780  Ставропольский край, г. Ставрополь, пр. К. Мар...  41.976486   
228     42810  Тульская обл, г. Тула, Красноармейский пр., до...  37.610727   
239   2633928  Челябинская обл, г. Магнитогорск, пр. Ленина, ...  58.984873   

       Долгота        Дата открытия Дата закрытия  Площадь магазина  \
16   48.698107  2010-03-14 00:00:00           NaT               386   
72   59.935056  2005-03-03 00:00:00           NaT               234   
155  58.524624  2011-04-14 00:00:00           NaT               375   
159  55.031426  2018-04-18 00:00:00           NaT               430   
166  53.197092  2011-09-24 00:00:00           NaT               311   
221  45.046667  2001-12-02 00:00:00           NaT                91   
228  54.196778  2010-02-14 00:00:00           NaT               363   
239  53.420774  2007-03-09 00:00:00           NaT               320   

                  Тип входа  Площадь ТЦ Формат магазина             Парковка  \
16   отдельный вход с улицы         NaN           Стрит                  NaN   
72   отдельный вход с улицы         NaN           Стрит                  NaN   
155  отдельный вход с улицы         NaN           Стрит  бесплатная парковка   
159   без сквозного прохода         NaN           Стрит  бесплатная парковка   
166  отдельный вход с улицы         NaN           Стрит                  NaN   
221  отдельный вход с улицы         NaN           Стрит                  NaN   
228  отдельный вход с улицы         NaN           Стрит                  NaN   
239  отдельный вход с улицы         NaN           Стрит                  NaN   

    Этаж      Время работы  Заполненность ТЦ  Рядом с эскалатором  \
16     1  с 10:00 до 21:00               3.0                    1   
72     1  с 10:00 до 22:00               NaN                    0   
155    3  с 10:00 до 21:00               NaN                    0   
159    1  с 10:00 до 22:00               2.0                    1   
166    1  с 10:00 до 21:00               3.0                    0   
221    1  с 10:00 до 21:00               NaN                    0   
228    1  с 10:00 до 21:00               3.0                    1   
239    1  с 10:00 до 21:00               NaN                    1   

     Рядом со входом  Конкуренты в ТЦ  Расположение в ТЦ     Толока  \
16                 0              NaN                NaN  19.174899   
72                 1              NaN                3.0  77.000192   
155                1              NaN                NaN  21.730924   
159                0              NaN                NaN  25.126273   
166                0              NaN                NaN  27.157986   
221                0              0.0                NaN  24.288103   
228                0              NaN                NaN  26.065731   
239                1              NaN                NaN  16.728042   

          Люкс   Выборка  
16   69.241379  Тестовая  
72   77.965517  Тестовая  
155  48.931034  Тестовая  
159  76.937500  Тестовая  
166  63.206897  Тестовая  
221  68.103448  Тестовая  
228  60.068966  Тестовая  
239  54.103448  Тестовая

In [36]:
len(street_shops)

8

## Число магазинов формата стрит в тестовой выборке:

Получено `8` магазинов формата стрит из тестовой выборки. 

### Уточнение по данным:
У точки с `id=42059` присутствует параметр `Расположение в ТЦ` как и `Заполненность ТЦ` у большенства других семплов, что требует уточнение у клиента.
Точка `42466` имеет в названии аббривиатуру `ТРК` и находится на `3` этаже, при этом указано, что `отдельный вход с улицы` это тоже требуется уточнить у клиента, для определения являются ли данные точки подходящими.

Определим `id` мини ТЦ, которые открылись до 2017 года

In [37]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 245 entries, 0 to 245
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   id точки             245 non-null    int64         
 1   Адрес                245 non-null    object        
 2   Широта               241 non-null    float64       
 3   Долгота              241 non-null    float64       
 4   Дата открытия        244 non-null    object        
 5   Дата закрытия        9 non-null      datetime64[ns]
 6   Площадь магазина     245 non-null    int64         
 7   Тип входа            243 non-null    object        
 8   Площадь ТЦ           203 non-null    float64       
 9   Формат магазина      245 non-null    object        
 10  Парковка             225 non-null    object        
 11  Этаж                 240 non-null    object        
 12  Время работы         231 non-null    object        
 13  Заполненность ТЦ     181 non-null  

In [38]:
df1['Дата открытия'] = pd.to_datetime(df1['Дата открытия'])

/var/folders/s5/t6cm63bn7zn0tzshtl9dsnm80000gn/T/ipykernel_94279/904020567.py:1: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  df1['Дата открытия'] = pd.to_datetime(df1['Дата открытия'])


In [39]:
valid_mini_malls=df1.loc[df1['Дата открытия'].dt.year < 2017]
valid_mini_malls.loc[valid_mini_malls['Формат магазина'] == 'Мини ТЦ']

id точки                                              Адрес      Широта  \
18      18331  Вологодская обл, г. Череповец, Советский пр., ...   37.930204   
24      18728       г. Вологда, ул. Благовещенская, дом № 4, ЦУМ   39.884911   
30      41824  г. Москва, Комсомольский пр., дом № 24, стр. 1...   37.580360   
47      41933  г. Москва, ул. Краснодарская, дом № 51, корпус...   37.765400   
66      42039  г. Санкт-Петербург, Ленинский пр., дом № 128, ...   30.265561   
73      42067  г. Санкт-Петербург, Невский пр., дом № 48, ТД ...   30.334499   
82      42113  г. Санкт-Петербург, пр. Просвещения, дом № 84,...   30.414297   
94      42163  г. Санкт-Петербург, ул. Одоевского, дом № 29, ...   30.234984   
99      42192  г. Санкт-Петербург,ул. Ленсовета, дом № 101, Л...   30.348522   
105     42219  Калужская обл, г. Калуга, ул. Кирова, дом № 20...   36.249798   
112     42242  Краснодарский край, Анапский р-он, г. Анапа, у...   37.321398   
117     42276  Краснодарский край, г. Сочи, Лесной пер., дом ...   39.723574   
118     42278  Краснодарский край, г. Сочи, п. Лазаревское, у...   39.335924   
131     42355  Ленинградская обл, г. Пушкин, ул. Московская, ...   30.405942   
133     42368  Ленинградская обл, Гатчинский р-он, г. Гатчина...   30.122856   
136     42383  М.О. г. Мытищи, Шараповский проезд, 1, "ТЦ "Кр...   37.761646   
145     42422  Московская обл, Раменский р-он, г. Раменское, ...   38.228595   
152     42458  Мурманская обл, г. Мурманск, ул. Ленина, дом №...   33.074807   
163     42496    Омская обл, г. Омск, ул. Масленникова, дом № 28   73.386644   
165     42504  Оренбургская обл, г. Оренбург, ул. Ленинская/П...   55.104088   
174     42548  Респ. Бурятия, г. Улан-Удэ, ул. Ленина, дом № ...  107.584862   
184     42618  Респ. Татарстан, г. Набережные Челны, бул. Энт...   52.421269   
186     42645  Респ. Татарстан, г. Нижнекамск, ул. Химиков, д...   51.814376   
195     42683     Россия, Пермский край, Пермь, улица Попова, 16   56.228132   
204     42720  Ростовская обл, г. Ростов-на-Дону, ул. Большая...   39.700000   
211     42743  Самарская обл, г. Самара, ул. Куйбышева, дом №...   50.088883   
227     42803  Томская обл, г. Томск, ул. 1905 года, дом № 14...   84.949903   
238     42645  Респ. Татарстан, г. Нижнекамск, ул. Химиков, д...   51.814376   
242   2633930  Ямало-Ненецкий АО, Ноябрьск г, Советская ул, д...   75.465427   

       Долгота Дата открытия Дата закрытия  Площадь магазина  \
18   59.125277    2007-02-21           NaT               234   
24   59.220169    2010-07-13           NaT               189   
30   55.728076    2009-10-09           NaT               265   
47   55.677300    2012-07-03           NaT               507   
66   59.852265    2006-09-25           NaT               148   
73   59.935376    2008-08-02           NaT               242   
82   60.037104    2008-03-14           NaT               297   
94   59.948167    2007-04-03           NaT               110   
99   59.831734    2012-12-23    2017-08-01               321   
105  54.514880    2012-11-25           NaT               545   
112  44.901369    2011-04-29           NaT               420   
117  43.593275    2008-09-10           NaT               193   
118  43.911720    2013-12-15           NaT               356   
131  59.719542    2005-04-24           NaT               239   
133  59.565871    2013-11-11           NaT               248   
136  55.914736    2010-07-12    2017-07-22               236   
145  55.566122    2011-05-07           NaT               192   
152  68.968507    2006-11-21           NaT               223   
163  54.972687    2012-04-27           NaT               298   
165  51.761131    2013-12-22           NaT               391   
174  51.827084    2012-08-11           NaT               247   
184  55.744911    2004-07-02           NaT               215   
186  55.631285    2006-04-16           NaT               255   
195  58.008433    2012-07-15           NaT               277   
204 

In [40]:
valid_mini_malls = valid_mini_malls['id точки'].loc[valid_mini_malls['Формат магазина'] == 'Мини ТЦ']

In [41]:
valid_mall_ids = list(set(valid_mini_malls).intersection(df2['id точки']))
valid_mall_ids

[42368,
 42496,
 42242,
 42504,
 42383,
 18728,
 42803,
 42548,
 42422,
 42683,
 41933,
 42192,
 42067,
 42458,
 41824,
 42720,
 42219,
 42355,
 42743]

Проверим полученные `id` еще раз на соответствие `Мини ТЦ` в исходном датасете

In [42]:
df1['Формат магазина'].loc[df1['id точки'].isin(valid_mall_ids)].unique()

array(['Мини ТЦ', 'ТЦ'], dtype=object)

In [43]:
df1.loc[df1['id точки'].isin(valid_mall_ids)].loc[df1['Формат магазина'] == 'ТЦ']

id точки                                              Адрес     Широта  \
194     42683  Россия, Пензенская область, Пенза, проспект Ст...  44.951652   

       Долгота Дата открытия Дата закрытия  Площадь магазина  \
194  53.220492    2013-12-27           NaT               578   

                 Тип входа  Площадь ТЦ Формат магазина             Парковка  \
194  без сквозного прохода     54500.0              ТЦ  бесплатная парковка   

    Этаж      Время работы  Заполненность ТЦ  Рядом с эскалатором  \
194    1  с 10:00 до 22:00               3.0                    1   

     Рядом со входом  Конкуренты в ТЦ  Расположение в ТЦ     Толока  Люкс  \
194                0              1.0                8.0  34.480876  60.5   

       Выборка  
194  Обучающая

In [44]:
df1.loc[df1['id точки'] == 42683]

id точки                                              Адрес     Широта  \
194     42683  Россия, Пензенская область, Пенза, проспект Ст...  44.951652   
195     42683     Россия, Пермский край, Пермь, улица Попова, 16  56.228132   

       Долгота Дата открытия Дата закрытия  Площадь магазина  \
194  53.220492    2013-12-27           NaT               578   
195  58.008433    2012-07-15           NaT               277   

                 Тип входа  Площадь ТЦ Формат магазина             Парковка  \
194  без сквозного прохода     54500.0              ТЦ  бесплатная парковка   
195  без сквозного прохода      6400.0         Мини ТЦ         нет парковки   

    Этаж      Время работы  Заполненность ТЦ  Рядом с эскалатором  \
194    1  с 10:00 до 22:00               3.0                    1   
195    1  с 10:00 до 21:00               NaN                    0   

     Рядом со входом  Конкуренты в ТЦ  Расположение в ТЦ     Толока  \
194                0              1.0                8.0  34.480876   
195                0              NaN                8.0  33.196617   

          Люкс    Выборка  
194  60.500000  Обучающая  
195  63.896552  Обучающая

### Замечание:
В исходном датасете `df1`, исходя из выше результатов, имеются дублирующиеся `id точка`(42683), соответвенно, использование таких данных для расчета средней выручки не возможно, поскольку данные о выручке по обучающей выборке используют `id точка`, как внешний ключ для сопоставления данных. Из основного датафрейма `df1` не исключаем, тк по всем остальным параметрам это разные магазины, один из которых имеет бесплатную парковку в будет в включен в ответ на 3 вопрос

In [45]:
valid_mall_ids = [i for i in valid_mall_ids if i != 42683]

In [46]:
mini_malls_profit = df2.loc[df1['id точки'].isin(valid_mall_ids)]
mini_malls_profit = mini_malls_profit[mini_malls_profit.columns[1:]]
mini_malls_profit

2015-01-01     2015-02-01     2015-03-01     2015-04-01  \
24   347460.172000  384190.158929  493988.175806  378390.484333   
30   583088.329380  697080.365802  866877.763798  643916.365400   
47   271799.961667  302923.795883  352652.033226  290865.504333   
73   391121.247000  527641.290000  621516.347419  435780.807000   
99   172584.289032  219756.954286  276604.092903  177366.738000   
105  398229.667896  500531.586762  580056.316452  416727.892004   
112   80410.251667  104142.941429  185292.591613  132366.491647   
131            NaN            NaN            NaN            NaN   
133            NaN            NaN            NaN            NaN   
136            NaN            NaN            NaN            NaN   
145            NaN            NaN            NaN            NaN   
152  378429.753517  350607.477988  482925.193552  367225.245935   

        2015-05-01     2015-06-01     2015-07-01     2015-08-01  \
24   337814.909677  346342.849333  317172.700645  338366.836452   
30   596320.837494  611306.690667  520712.425660  615215.181052   
47   244778.111290  259172.729277  261329.328710  282343.830968   
73   362702.244194  449316.747000  394088.948710  403953.500323   
99   158834.874667  197529.253333  174931.129032  208360.836129   
105  396598.559936  474847.306926  533608.867899  463919.701935   
112  159960.071616  137601.386968  140538.481240  178030.733548   
131            NaN            NaN            NaN            NaN   
133            NaN            NaN            NaN            NaN   
136            NaN            NaN            NaN            NaN   
145            NaN            NaN            NaN            NaN   
152  349294.028363  388179.193333  400668.150126  421847.743068   

        2015-09-01     2015-10-01     2015-11-01      2015-12-01  \
24   355564.976000  354175.949619  423315.924333   863454.330381   
30   623097.925336  635630.486452  670334.836667  1515944.004781   
47   288790.117000  305781.317419  317876.194333   561689.834733   
73   454876.422000  509798.104839     479802.609  1009995.189997   
99   200176.081379  219909.610323     214210.076   445304.384748   
105  519410.551000  529441.748652  496340.421996  1027985.676348   
112  167024.817333  221317.859677  179562.972667   396094.213813   
131            NaN            NaN            NaN             NaN   
133            NaN            NaN            NaN             NaN   
136            NaN            NaN            NaN             NaN   
145            NaN            NaN            NaN             NaN   
152  345782.485333  360541.050968     421239.038   957476.388168   

        2016-01-01     2016-02-01     2016-03-01  ...     2017-08-01  \
24   369665.179667  459482.020345  559246.229677  ...  289507.023871   
30   609746.760667    850236.8432  977176.848411  ...  745785.326452   
47   268720.875253  291410.395517  360927.856129  ...  263860.801935   
73      373389.951  567719.152759  658588.845484  ...  393217.493226   
99   208697.489333  268136.578621  331250.030323  ...  241577.343226   
105     464360.767  575351.161034   714670.67129  ...  634268.071935   
112  174285.939032   171552.46069  227175.470645  ...  223809.070000   
131            NaN            NaN            NaN  ...  343826.265484   
133            NaN            NaN            NaN  ...  340294.943226   
136            NaN            NaN            NaN  ...  252579.296594   
145            NaN            NaN            NaN  ...            NaN   
152   397970.02129  426358.826897  543228.298065  ...  416868.805161   

        2017-09-01     2017-10-01     2017-11-01    2017-12-01     2018-01-01  \
24   302238.579000  281283.290323  331402.619000  6.378567e+05  314928.539577   
30   749940.679333  633488.923871  813724.974667  1.439578e+06  694312.732040   
47   279768.044667  252862.284516  264948.201000  4.370723e+05  253066.556780   
73   395609.412000  365004.856452  410649.525000  7.288022e+05  364881.229020   
99   246786.954000  218930.980000  26

Теперь неободимо отфильтровать лишние колонки, в получившемся датафрейме, чтобы упростить рассчет среднего значения выручки

In [47]:
mini_malls_profit.columns = pd.to_datetime(mini_malls_profit.columns)
mini_malls_profit = mini_malls_profit.loc[:, mini_malls_profit.columns.year == 2016]

In [48]:
mini_malls_profit = mini_malls_profit.dropna(how='all')
mini_malls_profit

2016-01-01     2016-02-01     2016-03-01     2016-04-01  \
24   369665.179667  459482.020345  559246.229677  381858.573667   
30   609746.760667    850236.8432  977176.848411  725649.230667   
47   268720.875253  291410.395517  360927.856129  305084.787667   
73      373389.951  567719.152759  658588.845484     461519.982   
99   208697.489333  268136.578621  331250.030323  218518.301333   
105     464360.767  575351.161034   714670.67129     614294.239   
112  174285.939032   171552.46069  227175.470645  210685.002333   
131            NaN            NaN            NaN            NaN   
133            NaN            NaN            NaN            NaN   
136            NaN            NaN            NaN            NaN   
152   397970.02129  426358.826897  543228.298065  404400.638667   

        2016-05-01     2016-06-01     2016-07-01     2016-08-01  \
24   375772.788387  344001.565667  308478.825484  330966.439677   
30   813332.968387  706033.467333  609300.995481  688867.720000   
47   243601.171288  267928.675667  256540.522958  269025.945481   
73   408183.514839  467672.997000  389967.323226  392120.828710   
99   207277.721935  222148.099667  232665.450968  236915.669677   
105  512266.969355  584217.049996  549661.902584  549500.630323   
112  167739.096129  244897.743000  215656.407742  217351.654839   
131            NaN            NaN  215270.055000  243769.680000   
133            NaN            NaN            NaN            NaN   
136            NaN            NaN            NaN            NaN   
152  425347.254194  496725.236667  444494.209677  482043.645161   

        2016-09-01     2016-10-01     2016-11-01      2016-12-01  
24   358258.002667  360084.875484  356233.363637   758862.670826  
30   708808.942667  763585.803223  752322.609253   1612672.45332  
47   304101.984667  299991.237414     272532.824   440136.069138  
73   439969.266000  468169.339355   444261.70866    855072.67741  
99   211914.458667  254999.308667  230982.218513   420625.293897  
105  543732.390000  629722.228065  520683.020224  1047341.937984  
112  184653.047667  234590.964194  172160.644323   338863.992277  
131  245091.280500  273550.668387   257034.62466       590248.65  
133            NaN  354279.297692  309164.567513   505416.614342  
136            NaN  341370.640000  230342.263069   442256.414052  
152  401535.550000  425327.852258  400319.495073    838820.48609

In [49]:
mini_malls_profit.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11 entries, 24 to 152
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   2016-01-01 00:00:00  8 non-null      object 
 1   2016-02-01 00:00:00  8 non-null      object 
 2   2016-03-01 00:00:00  8 non-null      object 
 3   2016-04-01 00:00:00  8 non-null      object 
 4   2016-05-01 00:00:00  8 non-null      object 
 5   2016-06-01 00:00:00  8 non-null      float64
 6   2016-07-01 00:00:00  9 non-null      float64
 7   2016-08-01 00:00:00  9 non-null      float64
 8   2016-09-01 00:00:00  9 non-null      float64
 9   2016-10-01 00:00:00  11 non-null     float64
 10  2016-11-01 00:00:00  11 non-null     object 
 11  2016-12-01 00:00:00  11 non-null     object 
dtypes: float64(5), object(7)
memory usage: 1.1+ KB


В полученном датафрейме видно, что у некоторых мини ТЦ отсутствует выручка в начале 2016, проверим есть ли среди них, открывшиеся в 2016 году.

In [50]:
df1.loc[(df1['id точки'].isin(valid_mall_ids) & df1['Дата открытия'].dt.year == 2016)]

Empty DataFrame
Columns: [id точки, Адрес, Широта, Долгота, Дата открытия, Дата закрытия, Площадь магазина, Тип входа, Площадь ТЦ, Формат магазина, Парковка, Этаж, Время работы, Заполненность ТЦ, Рядом с эскалатором, Рядом со входом, Конкуренты в ТЦ, Расположение в ТЦ, Толока, Люкс, Выборка]
Index: []

И если, те которые закрылись в 2016 году или ранее.

In [51]:
df1.loc[(df1['id точки'].isin(valid_mall_ids) & df1['Дата закрытия'].dt.year)]

id точки                                              Адрес     Широта  \
99      42192  г. Санкт-Петербург,ул. Ленсовета, дом № 101, Л...  30.348522   
136     42383  М.О. г. Мытищи, Шараповский проезд, 1, "ТЦ "Кр...  37.761646   

       Долгота Дата открытия Дата закрытия  Площадь магазина  \
99   59.831734    2012-12-23    2017-08-01               321   
136  55.914736    2010-07-12    2017-07-22               236   

                 Тип входа  Площадь ТЦ Формат магазина             Парковка  \
99         сквозной проход         NaN         Мини ТЦ  бесплатная парковка   
136  без сквозного прохода      2000.0         Мини ТЦ         нет парковки   

    Этаж      Время работы  Заполненность ТЦ  Рядом с эскалатором  \
99     1  с 10:00 до 22:00               3.0                    0   
136    1  с 10:00 до 22:00               3.0                    0   

     Рядом со входом  Конкуренты в ТЦ  Расположение в ТЦ     Толока  \
99                 1              NaN                9.0  33.427061   
136                1              NaN                8.0  29.458197   

          Люкс    Выборка  
99   63.620690  Обучающая  
136  63.551724  Обучающая

### Уточнение:
Среди представленных мини ТЦ нет таких которые были открыты в 2016. В таком случае, отсутствующие значения в датафрейме `mini_malls_profit` можно рассматривать как нулевую выручку в соотвествующих месяцах или же как отсутствующая информацию. В данном случае требуется уточнение у клиента.
Как подразумевается расчет средней выручки? Расчет для каждого мини ТЦ средней выручки относительно всех месяцев за 2016 или относительно суммы всех месяцев за 2016? Далее будет решен первый вариант (для решения вторым необходимо использовать `sum(axis=1)`).

## Вопрос 2. Какова средняя выручка магазинов формата мини ТЦ за 2016 год?

In [52]:
avg_mm_profit = mini_malls_profit.mean(axis=1)
avg_mm_profit

/var/folders/s5/t6cm63bn7zn0tzshtl9dsnm80000gn/T/ipykernel_94279/904894417.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  avg_mm_profit = mini_malls_profit.mean(axis=1)


24     340357.941796
30     695319.385741
47     279517.673237
73     431579.950858
99     231728.597529
105    571366.840194
112    219429.963488
131    244420.420972
133    354279.297692
136    341370.640000
152    450025.298753
dtype: float64

### Уточнение:
По умолчанию метод `mean` игнорирует значения `nan` и не учитывает при вычислении среднего значения

In [53]:
avg_mm_profit = avg_mm_profit.mean()
avg_mm_profit

378126.9100235731

### Замечание и Уточнение

Вопрос клиенту, в чем измеряется выручка(в тысячах, миллионах)? Как округлять значения(до копеек или рублей)?

## Средняя выручка мини ТЦ за 2016г
Учитывая сделанные мной замечания и уточнения получаем следующее значение:
`378126.9100235731`

## Вопрос 3. Сколько магазинов с бесплатной парковкой?
Рассмотрим уникальные значения в колонке с парковкой

In [54]:
parking_values = df1['Парковка'].unique()
parking_values

array(['бесплатная парковка', 100, nan, 'платная парковка',
       'нет парковки', 1000, 700, 'бесплатная паpковка'], dtype=object)

В данном случае видим ситуацию аналогичную предыдущим - латиница и кириллица смешаны в одном слове

In [55]:
re.search(r'[a-zA-Z]', parking_values[-1]) 

<re.Match object; span=(13, 14), match='p'>

In [56]:
df1['Парковка'].loc[df1['Парковка'] == parking_values[-1]] = parking_values[0]

/var/folders/s5/t6cm63bn7zn0tzshtl9dsnm80000gn/T/ipykernel_94279/1287113697.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Парковка'].loc[df1['Парковка'] == parking_values[-1]] = parking_values[0]


In [57]:
df1['Парковка'].unique()

array(['бесплатная парковка', 100, nan, 'платная парковка',
       'нет парковки', 1000, 700], dtype=object)

In [58]:
df1['Парковка'].loc[df1['Парковка'] == parking_values[0]].count()

157

## Число магазинов с бесплатной парковкой
Всего `157` магазинов с бесплатной парковкой, с учетом семпла с повторяющимся `id` (42683)


### Замечание:
Требуется уточнить у клиента, как интерпретировать `nan` в парковках(как не известно или же ее отсутствие) для дальнейшего кодирования данных

# Дополнительные замечания клиенту

### Доп замечания по датафрейму `Выручка по обучающие выборке`
Рассмотрим еще раз датасет

In [59]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 152 entries, 0 to 153
Data columns (total 47 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id точки    152 non-null    int64  
 1   2015-01-01  116 non-null    float64
 2   2015-02-01  116 non-null    float64
 3   2015-03-01  116 non-null    float64
 4   2015-04-01  116 non-null    float64
 5   2015-05-01  117 non-null    float64
 6   2015-06-01  117 non-null    float64
 7   2015-07-01  117 non-null    float64
 8   2015-08-01  118 non-null    float64
 9   2015-09-01  119 non-null    float64
 10  2015-10-01  119 non-null    float64
 11  2015-11-01  120 non-null    object 
 12  2015-12-01  126 non-null    object 
 13  2016-01-01  127 non-null    object 
 14  2016-02-01  130 non-null    object 
 15  2016-03-01  131 non-null    object 
 16  2016-04-01  131 non-null    object 
 17  2016-05-01  131 non-null    object 
 18  2016-06-01  129 non-null    float64
 19  2016-07-01  132 non-null    f

Как можно заметить имеется проблема с типом данных, требуется проверка всех колоник с типом данных `object` на наличие "мусора" и дальнейшая замена типов на `float` и унификация по точности после запятой.

In [60]:
df2.loc[~df2['id точки'].isin(df1['id точки'])]

id точки     2015-01-01     2015-02-01     2015-03-01     2015-04-01  \
115     45801            NaN            NaN            NaN            NaN   
149   2633921  132965.439655  169366.746429  206388.450000  137964.700000   
150   2634000  341009.722000  413489.085501  518426.278387  376071.505667   
151   2634008  270923.715000  404737.144643  470433.872222  401063.781000   
152   2634011  378429.753517  350607.477988  482925.193552  367225.245935   
153   2633936            NaN            NaN            NaN            NaN   

        2015-05-01     2015-06-01     2015-07-01     2015-08-01  \
115            NaN            NaN            NaN            NaN   
149  104022.609677  148099.078333  109004.348387  142252.233871   
150  357895.794516  505318.240505  458096.520286  516279.226774   
151  318507.450000  385579.165000  375166.269677  382299.377093   
152  349294.028363  388179.193333  400668.150126  421847.743068   
153  420226.940000  311592.339222  255945.113890  234998.024826   

        2015-09-01     2015-10-01     2015-11-01     2015-12-01  \
115   16133.600000  209725.058065  150740.837333   316499.46431   
149  142128.645000  165113.769355  146348.766667   371015.23771   
150  486665.455265  570820.219042  650161.509667  971408.852693   
151  413219.906524  470166.098710  440464.357333  935522.989586   
152  345782.485333  360541.050968     421239.038  957476.388168   
153  234465.065333  279630.782667  301462.536667  555602.024777   

        2016-01-01     2016-02-01  ...     2017-08-01     2017-09-01  \
115   157374.19871  190895.928276  ...  254516.206452  227490.330667   
149  154774.131034  199913.774138  ...  113500.790323  105040.740000   
150  485994.616333  621628.405862  ...  621351.320968  548635.336667   
151  391747.646667  516522.805172  ...  236094.362903  242776.017667   
152   397970.02129  426358.826897  ...  416868.805161  406099.434987   
153  278632.685333  320643.226207  ...  376385.295819  333933.464667   

        2017-10-01     2017-11-01     2017-12-01     2018-01-01  \
115  224540.092903  270445.504000  459736.768671  249946.461213   
149  102333.540323  113391.018333  232945.668435  101905.779310   
150  514921.088710  687005.536667  929888.122658  513382.351353   
151  204961.915161  197560.122333  337194.695961  162847.124247   
152  369399.431613  467659.509333  887570.697890  461032.727239   
153  344517.292903  372648.691333  677214.710652  360466.441020   

        2018-02-01     2018-04-01     2018-05-01     2018-06-01  \
115  291878.760000  217643.365333  184792.629677  238328.832000   
149  132421.992857   86721.779167   90059.050000  113034.895000   
150  590733.692857  480563.392000  457005.916452  541734.577000   
151  202552.822857  185037.349333  167999.870323  186120.868667   
152  443111.015436  410602.549333  379963.923226  457012.976000   
153  372836.555000  299080.189333  286674.025806  327415.778667   

        2018-07-01     2018-08-01     2018-09-01     2018-10-01     2018-11-01  
115  218318.998710  260616.051613  238226.949333  253439.669677  288315.776000  
149   76825.856452   97766.093548   86143.466667   91317.248387  115013.301667  
150  444286.811935  574573.538710  538086.317000  530747.271290  565599.127667  
151  175067.582903  174118.710968  174518.631333  189597.414516  226016.234333  
152  431740.369032  491072.692000  461658.858000  421671.512258  489829.227333  
153  305542.250323  345389.650968  349845.855333  361092.704994  363068.734000  

[6 rows x 47 columns]

Некоторые значения отсутсвуют в первом датасете, соответственно их невозможно использовать для обучения модели МО

### Доп замечания по датафрейму `спавочник точек`

In [61]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 245 entries, 0 to 245
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   id точки             245 non-null    int64         
 1   Адрес                245 non-null    object        
 2   Широта               241 non-null    float64       
 3   Долгота              241 non-null    float64       
 4   Дата открытия        244 non-null    datetime64[ns]
 5   Дата закрытия        9 non-null      datetime64[ns]
 6   Площадь магазина     245 non-null    int64         
 7   Тип входа            243 non-null    object        
 8   Площадь ТЦ           203 non-null    float64       
 9   Формат магазина      245 non-null    object        
 10  Парковка             225 non-null    object        
 11  Этаж                 240 non-null    object        
 12  Время работы         231 non-null    object        
 13  Заполненность ТЦ     181 non-null  

### Вопрос клиенту:
 Что означают колонки `Толока` и `Люкс`?

### Замечания
- Многие колонки требуют замены типа данных.
- Для колонок `Тип входа`, `Формат магазина`, `Этаж` следует применить one-hot encoding
- Для колонок `Заполненность ТЦ`, `Конкуренты ТЦ` и `Расположение в ТЦ` использовать частотное кодирование, тк имеются отсутствующие значения
- Для `Время работы` требуется изменить представление данных(в случае если эта колонка будет иметь высокую вариативность при решении задачи МО)
- Не смотря на наличие широты и долготы, было бы не плохо разбить колонку с адресом, и добавить колонку `Регион` и `Город`, чтобы включить в модель спроса зависимость от региональных цен
- Для широты и долготы не хватает некоторых значений, их требуется определить(это возможно из-за наличия адреса), тк закодировать данные значения не получится.

Далее дополнительно рассмотрим колонки, которые могут представляться количественно, но по какой-то причине имеют тип данных `object`

In [62]:
df1['Этаж'].unique()

array([3, 1, 'цоколь', nan, 2, 4.5, 20,
       datetime.datetime(2019, 5, 2, 0, 0)], dtype=object)

### Замечание
Этаж не может исчисляться во временном формате

In [63]:
len(df1['Время работы'].unique()) 

26

Для колонки `Время работы` можно использовать частотное кодирование, как более простая альтернатива представления данных

In [64]:
df1['Тип входа'].unique()

array(['без сквозного прохода', 'сквозной проход',
       'отдельный вход с улицы + сквозной проход',
       'отдельный вход с улицы', 'единственный вход в ТК', nan],
      dtype=object)

На следующем этапе необходимо нормализовать и стандартизовать данные, провести выборку и извлечение свойств из справочника точек для дальнейшего формирования модели спроса.